spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Container.scala

```diff
@@ -41,18 +42,92 @@ abstract class Container[A <: Activity : ClassTag,
     B <: Activity : ClassTag, T: ClassTag](
   implicit ev: TensorNumeric[T]) extends AbstractModule[A, B, T] {
 
+  private var parent: Container[A, B, T] = null
```

Comment:
Maybe parent type should not be Container[A, B, T], e.g. this container output a tensor, so B is tensor. You cannot guarantee its parent output a tensor, right? Maybe Container[_, _, T] is better?

Reply:
There's still some potential issue even use the parent ref. For an example

C1 -> C2

There's a Linear in C2. If we add a new layer into C1, the output shape of C1 maybe changed. So we need to recreate a new Linear layer as the weight shape is changed

Instead of import a parent ref here, I suggest we frozen the container after it is added into another container.

Reply:
yeah, the recreating `Linear` is the overhead for such solution, but I suppose gc would handle the memory cleaning. If we "frozen" the container, then the behavior is not compatible with the current style. we need to think about this seriously again.

Reply:
I'm afraid recreating may import many potential issues.

Reply:
We can frozee the container only if the shape infer is enabled, so it won't conflic with current style...